dll/win32/shell32/shelldesktop/CChangeNotifyServer.cpp

```diff
@@ -6,18 +6,502 @@
+
+    BOOL Contains(LPCWSTR pszPath, BOOL fDir) const;
+
+    void RenameItem(LPCWSTR pszItem1, LPCWSTR pszItem2, BOOL fDir);
+    void DeleteItem(LPCWSTR pszItem, BOOL fDir);
+
+protected:
+    SIZE_T m_count;
+    LPWSTR m_items[ANYSIZE_ARRAY];
+
+    DIRLIST()
+    {
+    }
+};
+
+DIRLIST::~DIRLIST()
+{
+    if (this)
+    {
+        for (DWORD i = 0; i < m_count; ++i)
+        {
+            free(m_items[i]);
+        }
+        free(this);
+    }
+}
+
+BOOL DIRLIST::Contains(LPCWSTR pszPath, BOOL fDir) const
+{
+    if (this == NULL)
+        return FALSE;
+
+    for (DWORD i = 0; i < m_count; ++i)
+    {
+        if (m_items[i] == NULL)
+            continue;
+
+        if (fDir)
+        {
+            if (m_items[i][0] != L'|')
+                continue;
+        }
+        else
+        {
+            if (m_items[i][0] != L'>')
+                continue;
+        }
+
+        if (lstrcmpiW(&m_items[i][1], pszPath) == 0)
+            return TRUE;
+    }
+
+    return FALSE;
+}
+
+/*static*/ DIRLIST *
+DIRLIST::AddItem(DIRLIST *pList, LPCWSTR pszItem, BOOL fDir)
+{
+    SIZE_T count = 0, cbDirList = sizeof(DIRLIST);
+    if (pList)
+    {
+        count = pList->m_count;
+        cbDirList += count * sizeof(LPWSTR);
+    }
+
+    DIRLIST *pNewList = (DIRLIST *)realloc(pList, cbDirList);
+    if (pNewList == NULL)
+        return pList;
+
+    if (count == 0)
+        ZeroMemory(pNewList, cbDirList);
+
+    WCHAR szPath[MAX_PATH + 1];
+    szPath[0] = fDir ? L'|' : L'>';
+    lstrcpynW(&szPath[1], pszItem, _countof(szPath) - 1);
+
+    pNewList->m_items[pNewList->m_count] = _wcsdup(szPath);
+    pNewList->m_count++;
+    return pNewList;
+}
+
+void DIRLIST::RenameItem(LPCWSTR pszItem1, LPCWSTR pszItem2, BOOL fDir)
+{
+    if (this == NULL)
+        return;
+
+    WCHAR szPath[MAX_PATH + 1];
+    for (DWORD i = 0; i < m_count; ++i)
+    {
+        if (m_items[i] && lstrcmpiW(&m_items[i][1], pszItem1) == 0)
+        {
+            szPath[0] = fDir ? L'|' : L'>';
+            lstrcpynW(&szPath[1], pszItem2, _countof(szPath) - 1);
+
+            free(m_items[i]);
+            m_items[i] = _wcsdup(szPath);
+            return;
+        }
+    }
+}
+
+void DIRLIST::DeleteItem(LPCWSTR pszItem, BOOL fDir)
+{
+    if (this == NULL)
+        return;
+
+    for (DWORD i = 0; i < m_count; ++i)
+    {
+        if (m_items[i] && lstrcmpiW(&m_items[i][1], pszItem) == 0)
+        {
+            free(m_items[i]);
+            m_items[i] = NULL;
+            return;
+        }
+    }
+}
+
+/*static*/ DIRLIST *
+DIRLIST::GetDirList(DIRLIST *pList, LPCWSTR pszDir, BOOL fRecursive)
+{
+    WCHAR szPath[MAX_PATH];
+    GetFullPathNameW(pszDir, _countof(szPath), szPath, NULL);
+
+    if (!PathIsDirectoryW(szPath) && !PathIsRootW(szPath))
+    {
+        delete pList;
+        return NULL;
+    }
+
+    pList = AddItem(pList, szPath, TRUE);
+
+    WIN32_FIND_DATAW find;
+    PathAppendW(szPath, L"*");
+
+    HANDLE hFind = FindFirstFileW(szPath, &find);
```

Comment:
What is the purpose of this?

Reply:
To remember the file list of the directory.